ee/desktop/menu/menu.go

```diff
@@ -1,36 +1,196 @@
-		versionItem.Disable()
+const (
+	KolideDesktopIcon      = "kolide-desktop"
+	KolideDebugDesktopIcon = "kolide-debug-desktop"
+)
 
-		// if prod environment, return
-		if hostname == "k2device-preprod.kolide.com" || hostname == "k2device.kolide.com" {
-			return
-		}
+// MenuData encapsulates a menu bar icon and accessible menu items
+type MenuData struct {
+	Icon    menuIcon       `json:"icon"`
+	Tooltip string         `json:"tooltip,omitempty"`
+	Items   []menuItemData `json:"items"`
+}
+
+// menuItemData represents a menu item, optionally containing sub menu items
+type menuItemData struct {
+	Label       string         `json:"label,omitempty"`
+	Tooltip     string         `json:"tooltip,omitempty"`
+	Disabled    bool           `json:"disabled,omitempty"`
+	NonProdOnly bool           `json:"nonProdOnly,omitempty"`
+	IsSeparator bool           `json:"isSeparator,omitempty"`
+	Action      actionData     `json:"action,omitempty"`
+	Items       []menuItemData `json:"items,omitempty"`
+}
+
+// MenuBuilder is an interface a menu parser can use to specify how the menu is built
+type MenuBuilder interface {
+	// SetIcon sets the menu bar icon
+	SetIcon(icon menuIcon)
+	// SetTooltip sets the menu tooltip
+	SetTooltip(tooltip string)
+	// AddMenuItem creates a menu item with the supplied attributes. If the menu item is successfully
+	// created, it is returned. If parent is non-nil, the menu item will be created as a child of parent.
+	AddMenuItem(label, tooltip string, disabled, nonProdOnly bool, action Action, parent any) any
+	// AddSeparator adds a separator to the menu
+	AddSeparator()
+}
+
+type menu struct {
+	logger     log.Logger
+	hostname   string
+	filePath   string
+	buildMutex sync.Mutex
+	doneChans  []chan<- struct{}
+}
+
+func New(logger log.Logger, hostname, filePath string) *menu {
+	m := &menu{
+		logger:   logger,
+		hostname: hostname,
+		filePath: filePath,
+	}
 
-		// in non prod environment
+	return m
+}
+
+// Init creates the menu bar & icon. It must be called on the main thread, and
+// blocks until Shutdown() is called.
+func (m *menu) Init() {
+	// Build will be invoked after the menu has been initialized
+	// Before the menu exits, cleanup the goroutines
+	systray.Run(m.Build, m.cleanup)
+}
+
+// Build parses the menu file and constructs the menu. If a menu already exists,
+// all of its items will be removed before the new menu is built.
+func (m *menu) Build() {
+	// Lock so the menu is never being modified by more than one goroutine at a time
+	m.buildMutex.Lock()
+	defer m.buildMutex.Unlock()
+
+	// Remove all menu items each time we rebuild the menu
+	systray.ResetMenu()
+
+	// Even though the menu items have been removed, we still have goroutines hanging around
+	m.cleanup()
+
+	// Reparse the menu file & rebuild the menu
+	menuData := m.getMenuData()
+	if menuData == nil {
+		var defaultMenuData MenuData
+		menuData = &defaultMenuData
+	}
```

Comment:
Is the intention here just to set menuData to the zero value?

How about:
```suggestion
	if menuData == nil {
		menuData = &MenuData{}
	}
```

Reply:
Fixed this, now just falling back to `getDefaultMenu()`